More software engineering codes are:

---------- 2008_C_R_MIMAR_PHD
MIMAR package was shared to the population genetic community, well documented (see pdf files) and maintained until 2010.
-mimardir is the main C program that estimates the demographic parameters of an "isolation-migration" model from recombining loci.
-mimargofdir: 
			Perl scripts perlgofest and perlgof call the program MIMARgof to generate samples under an "isolation-migration" model with either point estimates of the parameters provided by MIMAR, or sets of parameters sampled from the posterior distribution estimated by MIMAR
			+ R script "testGOF.R" to help the user test whether the estimated model fit the data.
-mimarsimdir: C code that simulates a data set under the "isolation- migration" model 


--------------2012_py_R_java_BaseHealth/algogene-pipeline
AlgoGenePipeline.py <Disease_name> <lifetime_prevalence>: 
   I)   variant aggregation : all the information is gathered for the genetic variation of interest
   II)  Data aggregation    : all the variants are aggregated for the disease of interest
   III) Data selection      : good variants are selected for the disease of interest (QC & selection routine)
   IV)  Adjusted calculation: selected variatns go throug adjusted odds ration calculation
   V)   Ethnicity decision  : Genophen covers about 6 different ethinicities at the time of this writting. what is the default ethnicity for the disease of interest


-------------- 2016_py_get_google_pmids
gets lists of pmids from google sheets in specific google drive directories

-------------- 2016_py_InvitaeCode
Translates a (0-based) transcript coordinate to a (0-based) genome coordinate.
https://rawgit.com/celinesf/personal/master/2016_py_InvitaeCode/doc/html/index.html
see documentation [here](https://rawgit.com/celinesf/personal/master/2016_py_InvitaeCode/doc/html/index.html)
see documentation <a href="https://rawgit.com/celinesf/personal/master/2016_py_InvitaeCode/doc/html/index.html" target="_blank"><b>here</b></a>
